### DISCUSSION TITLE
Factory package aanmaken

### DISCUSSION THREAD

**Discussion by pjotrcasteel** [MatrTech/EFCore#2] [Ideas]
Wij moeten de generic factory packagebaar maken

**Comment by visschersm** [comment]
Welke generic factory bedoel je hier mee?

**Comment by pjotrcasteel** [comment]
Voor het aanmaken van bijv, de context

**Comment by visschersm** [comment]
I think this was about the efcore dependency injection using service scope factory before dotnet 5. 
Might be a nice research item to see if we can implement a factory that can manage both the pre dotnet way and the new way using the DbContextFactory.

### ServiceScopeFactory registration:
```csharp
public void ConfigureServices(IServiceCollection services)
{
    services.AddControllers();

    services.AddDbContext<ApplicationDbContext>(
        options => options.UseSqlServer("name=ConnectionStrings:DefaultConnection"));
}
```

### ServiceScopeFactory injection:
```csharp
private readonly IServiceScopeFactory  scopeFactory;

public MyController(IServiceScopeFactory   scopeFactory)
{
    this.scopeFactory = scopeFactory;
}
```

### ServiceScopeFactory usage:
```csharp
public void DoSomething()
{
    using(var scope = serviceScopeFactory.CreateScope())
    {
        var context = scope.ServiceProvider.GetService<MyDataContext>();
    
        // ...
    }
}
```

### DbContextFactory registration:
```csharp
public void ConfigureServices(IServiceCollection services)
{
    services.AddControllers();

    services.AddDbContextFactory<ApplicationDbContext>(
        options =>
            options.UseSqlServer(@"Server=(localdb)\mssqllocaldb;Database=Test"));
}
```

### DbContextFactory injection:
```csharp
private readonly IDbContextFactory<ApplicationDbContext> _contextFactory;

public MyController(IDbContextFactory<ApplicationDbContext> contextFactory)
{
    _contextFactory = contextFactory;
}
```

### DbContextFactory usage:
```csharp
public void DoSomething()
{
    using (var context = _contextFactory.CreateDbContext())
    {
        // ...
    }
}
```

[DbContextFactory reference](https://docs.microsoft.com/en-us/ef/core/dbcontext-configuration/#using-a-dbcontext-factory-eg-for-blazor)
[ServiceScopeFactory reference](https://blog.hildenco.com/2018/12/accessing-entity-framework-context-on.html)

**Comment by visschersm** [comment]
It might also connect nicely to our [GenericFactory](https://github.com/MatrTech/Utilities.Test/blob/main/src/Utilities.Test/GenericFactory.cs) 
Lets start a discussion on how these packages could work together.

[How would this map to the EFCore usecase](https://github.com/MatrTech/Utilities.Test#mstest-2)
